torch/_tensor.py

```diff
@@ -908,6 +908,40 @@ def rename(self, *names, **rename_map):
         # See Note [rename_ / rename API]
         return update_names(self, names, rename_map, inplace=False)
 
+    def to_sparse_csr(self):
+        """ Convert a tensor to compressed row storage format. Only works with 2D tensors.
+
+        Examples::
+
+            >>> dense = torch.randn(5, 5)
+            >>> sparse = dense.to_sparse_csr()
+            >>> sparse._nnz()
+            25
+
+        """
+        shape = self.size()
+        fill_value = 0
+        if len(shape) != 2:
+            raise RuntimeError("Only 2D tensors can be converted to the CSR format but got shape: ", shape)
+
+        if self.is_sparse:
+            coalesced_self = self.coalesce()
+            row_indices = coalesced_self.indices()[0]
+            ro = [0]
+            i = 0
+            for irow in range(self.shape[0]):
+                while i < row_indices.size()[0] and row_indices[i] == irow:
+                    i += 1
+                ro.append(i)
```

Comment:
This looks... very slow

Reply:
Considering that we need to read each element in `row_indices` at least once and that this algorithm would increment `i` from 0 to `len(row_indices)`, isn't an O(N) time alright?

Reply:
It looks very slow not because of asymptotics, but because you're running the algorithm in Python

Reply:
Resolving via https://github.com/pytorch/pytorch/issues/57381